src/persist/src/indexed/columnar/arrow.rs

```diff
@@ -117,6 +98,74 @@ pub(crate) const ENABLE_ARROW_LGALLOC_NONCC_SIZES: Config<bool> = Config::new(
     "A feature flag to enable copying decoded arrow data into lgalloc on non-cc sized clusters.",
 );
 
+fn realloc_data(data: ArrayData, metrics: &ColumnarMetrics) -> ArrayData {
+    // NB: Arrow generally aligns buffers very coarsely: see arrow::alloc::ALIGNMENT.
+    // However, lgalloc aligns buffers even more coarsely - to the page boundary -
+    // so we never expect alignment issues in practice. If that changes, build()
+    // will return an error below, as it does for all invalid data.
+    let buffers = data
+        .buffers()
+        .iter()
+        .map(|b| realloc_buffer(b, metrics))
+        .collect();
+    let child_data = data
+        .child_data()
+        .iter()
+        .cloned()
+        .map(|d| realloc_data(d, metrics))
+        .collect();
+    let nulls = data.nulls().map(|n| {
+        let buffer = realloc_buffer(n.buffer(), metrics);
+        NullBuffer::new(BooleanBuffer::new(buffer, n.offset(), n.len()))
+    });
+
+    data.into_builder()
+        .buffers(buffers)
+        .child_data(child_data)
+        .nulls(nulls)
+        .build()
+        .expect("reconstructing valid arrow array")
+}
+
+/// Re-allocate the backing storage for a specific array using lgalloc, if it's configured.
+pub fn realloc_array<A: Array + From<ArrayData>>(array: &A, metrics: &ColumnarMetrics) -> A {
+    let data = array.to_data();
+    let data = realloc_data(data, metrics);
+    A::from(data)
+}
+
+/// Re-allocate the backing storage for an array ref using lgalloc, if it's configured.
+pub fn realloc_any(array: &ArrayRef, metrics: &ColumnarMetrics) -> ArrayRef {
+    let data = array.to_data();
+    let data = realloc_data(data, metrics);
+    make_array(data)
+}
+
+fn realloc_buffer(buffer: &Buffer, metrics: &ColumnarMetrics) -> Buffer {
+    let use_lgbytes_mmap = if metrics.is_cc_active {
+        ENABLE_ARROW_LGALLOC_CC_SIZES.get(&metrics.cfg)
+    } else {
+        ENABLE_ARROW_LGALLOC_NONCC_SIZES.get(&metrics.cfg)
+    };
+    let region = if use_lgbytes_mmap {
+        metrics
+            .lgbytes_arrow
+            .try_mmap_region(buffer.as_slice())
+            .ok()
```

Comment:
(Not actionable right now) I'd like to move us away from `Region` to a better abstraction which presents as a fixed-length array instead of something that kinda looks like a vector with strange semantics. Just pointing it out here so everyone's aware the `Region` type should eventually disappear!

Reply:
Makes sense! Here all we need is a fixed-length array, so it should be straightforward to move over when the time comes...